src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,202 @@
+
+        restore_reg     METHODDESC_REGISTER, UMThunkStubAMD64_METHODDESC_OFFSET
+
+#if DEBUG
+        mov             rax, [r12 + OFFSETOF__Thread__m_pDomain]
+        mov             eax, [rax + OFFSETOF__AppDomain__m_dwId]
+
+        mov             r11d, [METHODDESC_REGISTER + OFFSETOF__UMEntryThunk__m_dwDomainId]
+
+        cmp             rax, r11
+        jne             LOCAL_LABEL(WrongAppDomain)
+#endif
+
+        mov             r11, [METHODDESC_REGISTER + OFFSETOF__UMEntryThunk__m_pUMThunkMarshInfo]
+        mov             eax, [r11 + OFFSETOF__UMThunkMarshInfo__m_cbActualArgSize]                      // stack_args
+        test            rax, rax                                                                        // stack_args
+        jnz             LOCAL_LABEL(UMThunkStub_CopyStackArgs)                                          // stack_args
+        
+LOCAL_LABEL(UMThunkStub_ArgumentsSetup):
+        mov    rdi, [rbp + UMThunkStubAMD64_INT_ARG_OFFSET]
+        mov    rsi, [rbp + UMThunkStubAMD64_INT_ARG_OFFSET + 0x08]
+        mov    rdx, [rbp + UMThunkStubAMD64_INT_ARG_OFFSET + 0x10]
+        mov    rcx, [rbp + UMThunkStubAMD64_INT_ARG_OFFSET + 0x18]
+        mov    r8,  [rbp + UMThunkStubAMD64_INT_ARG_OFFSET + 0x20]
+        mov    r9,  [rbp + UMThunkStubAMD64_INT_ARG_OFFSET + 0x28]
+        movdqa xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET]
+        movdqa xmm1, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x10]
+        movdqa xmm2, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x20]
+        movdqa xmm3, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x30]
+        movdqa xmm4, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x40]
+        movdqa xmm5, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x50]
+        movdqa xmm6, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x60]
+        movdqa xmm7, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 0x70]
+
+        mov             rax, [r11 + OFFSETOF__UMThunkMarshInfo__m_pILStub]                              // rax <- Stub*
+        call            rax
+
+LOCAL_LABEL(PostCall):
+        //
+        // enable preemptive GC
+        //
+        mov             dword ptr [r12 + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 0
+
+        // epilog
+        lea             rsp, [rbp + UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE]
+        pop_nonvol_reg  rbp                                                                             // stack_args
+        pop_nonvol_reg  r12
+        ret
+
+
+LOCAL_LABEL(DoThreadSetup):
+        call            C_FUNC(CreateThreadBlockThrow)
+        jmp             LOCAL_LABEL(HaveThread)
+        
+LOCAL_LABEL(DoTrapReturningThreadsTHROW):
+        restore_reg     METHODDESC_REGISTER, UMThunkStubAMD64_METHODDESC_OFFSET
+        mov             rdi, r12                                                                  // Thread* pThread
+        mov             rsi, METHODDESC_REGISTER                                                  // UMEntryThunk* pUMEntry
```

Comment:
This could be just: mov rsi, [rbp + UMThunkStubAMD64_METHODDESC_OFFSET]; and the restore_reg won't be necessary.